include/envoy/stats/stats.h

```diff
@@ -20,39 +21,92 @@ class Instance;
 namespace Stats {
 
 /**
+ * General representation of a tag.
+ */
+struct Tag {
+  std::string name_;
+  std::string value_;
+};
+
+class TagExtractor {
+public:
+  virtual ~TagExtractor() {}
+
+  /**
+   * Identifier for this tag.
+   */
+  virtual std::string name() const PURE;
+
+  /**
+   * Updates the tag extracted name and the set of tags by extracting the tag represented by this
+   * TagExtractor. If the tag is not represented in the current tag_extracted_name, nothing will be
+   * modified.
+   * @param tag_extracted_name name from which the tag will be removed if found to be
+   * represented in the name.
+   * @param tags list of tags updated with the tag name and value if found in the
```

Comment:
You might want to be extra clear here that you are concatenating and preserving existing list.